Add macOS support; input validation; an example

l50 opened the pull request:
This pull request contains the following:

- The program now works on macOS
- Included brew formulae and instructions to install
  - You will want to change the values in the rb file to match your release, sha, etc.
- There is input validation in place to stop the program from running when something isn't working
- Added missing input validation to ansible example
- Changed the `Makefile` values to relative paths from the build directory
- Created an example and provided documentation in the README

l50 committed:
Add macOS support and other improvements

- The program now works on macOS
- There is input validation in place to stop the program from running when something isn't working
- Changed the Makefile values to relative paths from the build directory

l50 committed:
Add example

l50 committed:
Updated headers for all scripts

kriansa commented on README.md:
```diff
@@ -128,3 +128,19 @@ Please make sure to update tests as appropriate. For more information, please re
 This project is licensed under the BSD 3-Clause License - see the [LICENSE.md](LICENSE.md) file for
 details.
+
+## Example
```
I think that by this instruction you meant to help people how to use this software without needing to install its package. I would recommend against it for the reasons I specified above.

On the other hand, you could leverage this text to actually instruct how to use it for developing purposes (i.e. you are developing ansible-bundler and you are trying to test it out)

l50 replied:
Added a developer section to the README, closing.

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -194,20 +194,20 @@ copy_playbook() {
 
 	echo "Adding playbook..."
-	cp -L --preserve=timestamps "$playbook_file" "$tmpdir/playbook.yml"
+	cp -L -p "$playbook_file" "$tmpdir/playbook.yml"
```
```suggestion
	cp -Lp "$playbook_file" "$tmpdir/playbook.yml"
```

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -194,20 +194,20 @@ copy_playbook() {
 	if [ -n "$requirements_file" ]; then
 		echo "Adding requirements..."
-		cp -L --preserve=timestamps "$requirements_file" "$tmpdir/requirements.yml"
+		cp -L -p "$requirements_file" "$tmpdir/requirements.yml"
```
```suggestion
		cp -Lp "$requirements_file" "$tmpdir/requirements.yml"
```

l50 replied:
Implemented project owner's preferred parameters for the `cp` command , closing.

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -194,20 +194,20 @@ copy_playbook() {
 	if [ -n "$vars_file" ]; then
 		echo "Adding vars..."
-		cp -L --preserve=timestamps "$vars_file" "$tmpdir/vars.yml"
+		cp -L -p "$vars_file" "$tmpdir/vars.yml"
```
```suggestion
		cp -Lp "$vars_file" "$tmpdir/vars.yml"
```

l50 replied:
Implemented project owner's preferred parameters for the `cp` command , closing.

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -218,7 +218,7 @@ copy_playbook() {
 
 	echo "Adding ansible config..."
-	cp -L --preserve=timestamps "$ETC_PATH/ansible.cfg" "$tmpdir"
+	cp -L -p "$ETC_PATH/ansible.cfg" "$tmpdir"
```
```suggestion
	cp -Lp "$ETC_PATH/ansible.cfg" "$tmpdir"
```

l50 replied:
Implemented project owner's preferred parameters for the `cp` command , closing.

kriansa commented on examples/basic.yaml:
```diff
@@ -0,0 +1,12 @@
+
+    - name: Print extra variable 'example'
+      debug:
+        msg: "Example value is {{ example }}"
```
Ensure the file ends with a newline. See `.editorconfig` for project defaults

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -290,20 +290,34 @@ install_roles() {
 
 compress_output() {
-	local output_file=$1
+    local output_file=$1
```
```suggestion
  local output_file=$1
```

Don't change the indentation. See `.editorconfig` for editor settings

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -290,20 +290,34 @@ install_roles() {
+    local output_file=$1
 
-	echo "Packaging files to the output..."
+    echo "Packaging files to the output..."
```
```suggestion
  echo "Packaging files to the output..."
```

Don't change the indentation. See .editorconfig for editor settings

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -290,20 +290,34 @@ install_roles() {
 	echo "Adding entrypoint..."
-	cp -L --preserve=timestamps "$LIB_PATH/run-playbook.sh" "$tmpdir"
-	chmod 0755 "$tmpdir/run-playbook.sh"
+    if [ -e "$LIB_PATH/run-playbook.sh" ]; then
```
This check should not be here, instead we must evaluate that `$LIB_PATH` is set to the correct path when this runs - and this is already done at `set_config_vars()`

l50 replied:
I'm afraid I don't understand your feedback here, but I'm happy to implement whatever proposed changes you have.

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -290,20 +290,34 @@ install_roles() {
 
-	# First add the header to the final script
-	cat "$LIB_PATH/bin-header.sh" > "$output_file"
+    # First add the header to the final script
```
This check should not be here, instead we must evaluate that `$LIB_PATH` is set to the correct path when this runs - and this is already done at `set_config_vars()`

l50 replied:
I'm afraid I don't understand your feedback here, but I'm happy to implement whatever proposed changes you have.

kriansa commented on app/bin/bundle-playbook:
```diff
@@ -290,20 +290,34 @@ install_roles() {
 	# Do some build-time metadata insertion
-	sed -i'' \
+	if [[ "$(uname)" == 'Darwin' ]]; then 
+		SED_BINARY=`which gsed`
```
Here you're adding a hidden dependency on gnu-sed which implies the user will have it installed.

I know that the only thing that kept you from using `sed` here was the fact that `-i` flag is incompatible between the platforms. We can easily solve that by:

```sh
sed -i.bkp -e '...' "$output_file"; rm "${output_file}.bkp"
```

And we avoid the conditional and the hidden dependency.

l50 replied:
Added docs for installing gnu grep if on a Mac to the README. It saves on the temp file.

kriansa commented on app/lib/bin-header.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env bash
```
Having it requiring `sh` was so that we add compatibility with OSes without bash pre-installed such as BSDs.

For portability let's keep the shebang invoking `sh` with `/usr/bin/env`:

```suggestion
#!/usr/bin/env sh
```

l50 replied:
I'd like to point out that the script employs some features specific to Bash and not just POSIX shell, including the use of `$(command -v ...)`. This Bash built-in is employed to check the availability of Python and pip:

```bash
if ! command -v python >/dev/null 2>&1 && ! command -v python3 >/dev/null 2>&1; then
```

Given that Bash is widely adopted and supports these features consistently, using a bash shebang helps ensure that the script behaves as expected across various environments. While some of these features might be present in other shells, the use of Bash-specific features in the script before my edits makes it reasonable to continue with the bash shebang.

If the author wishes to modify the rest of the scripts accordingly before this PR can be merged, they are welcome to do so. However, my suggestion is to maintain the bash shebang for the reasons outlined above.

kriansa commented on app/lib/run-playbook.sh:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env bash
```
Same as above:

```suggestion
#!/usr/bin/env sh
```

l50 replied:
I'd like to point out that the script employs some features specific to Bash and not just POSIX shell, including the use of $(command -v ...). This Bash built-in is employed to check the availability of Python and pip:

```bash
if ! command -v python >/dev/null 2>&1 && ! command -v python3 >/dev/null 2>&1; then
```

Given that Bash is widely adopted and supports these features consistently, using a bash shebang helps ensure that the script behaves as expected across various environments. While some of these features might be present in other shells, the use of Bash-specific features in the script before my edits makes it reasonable to continue with the bash shebang.

If the author wishes to modify the rest of the scripts accordingly before this PR can be merged, they are welcome to do so. However, my suggestion is to maintain the bash shebang for the reasons outlined above.

kriansa commented on app/lib/run-playbook.sh:
```diff
@@ -11,7 +11,11 @@ main() {
 
 	# Ensure we have HOME defined, otherwise set it manually
-	test -z "$HOME" && export HOME; HOME="$(getent passwd "$(id -un)" | cut -d: -f6)"
+	if [[ "$(uname)" == 'Darwin' ]]; then
```
`[[` is incompatible with `sh`

Keep `test` instead.

```suggestion
	if test "$(uname)" = "Darwin"; then
```

l50 replied:
I'd like to point out that the script employs some features specific to Bash and not just POSIX shell, including the use of $(command -v ...). This Bash built-in is employed to check the availability of Python and pip:

```bash
if ! command -v python >/dev/null 2>&1 && ! command -v python3 >/dev/null 2>&1; then
```

Given that Bash is widely adopted and supports these features consistently, using a bash shebang helps ensure that the script behaves as expected across various environments. While some of these features might be present in other shells, the use of Bash-specific features in the script before my edits makes it reasonable to continue with the bash shebang.

If the author wishes to modify the rest of the scripts accordingly before this PR can be merged, they are welcome to do so. However, my suggestion is to maintain the bash shebang for the reasons outlined above.

kriansa commented on app/lib/run-playbook.sh:
```diff
@@ -11,7 +11,11 @@ main() {
 	# Ensure we have HOME defined, otherwise set it manually
-	test -z "$HOME" && export HOME; HOME="$(getent passwd "$(id -un)" | cut -d: -f6)"
+	if [[ "$(uname)" == 'Darwin' ]]; then
+ 	   test -z "$HOME" && export HOME; HOME="$(dscl . -read /Users/$(id -un) NFSHomeDirectory | awk -F': ' '{print $2}')"
```
Since you're already relying on the path `/Users/$(id -un)` you don't need all of it.

It's quite possible that you can't have your home path in other directory under macOS anyway :man_shrugging: 

```suggestion
 	   test -z "$HOME" && export HOME; HOME="/Users/$(id -un)"
```

l50 replied:
Nice and concise, I'm a fan. Implemented.

kriansa commented on Makefile:
```diff
@@ -7,9 +7,9 @@ build_dir=build/pkg
-etc_path=/etc/ansible-bundler
+bin_path=.
+lib_path=../lib/ansible-bundler
+etc_path=../../etc/ansible-bundler
```
You don't need to change any of that, even if you just want to run `bin/bundle-playbook`, you don't have to build it.

Changing these will prevent the build for other platforms.

l50 replied:
Thank you for your feedback. I understand that changing these variables might prevent the build for other platforms. My intention was to avoid hardcoding the paths, as it could cause issues in different environments. However, I understand that it might not be the best solution for all cases.

Would you have any suggestions on how to make these paths more flexible without affecting other platforms? I am open to any ideas that could help us find a more suitable solution.

kriansa commented on Makefile:
```diff
@@ -32,7 +32,13 @@ package:
+	else \
+		SED_BINARY=$$(which sed); \
+	fi; \
+	$$SED_BINARY -i'' \
```
Just replace it with the `sed && rm` solution I pointed above.

l50 replied:
See the new changes I made with the brew formulae.

kriansa reviewed (changes requested):
Thanks for the contribution! It's really lovely to see that it's usable for macOS as well!

I understand that some of the changes you've made here are also to make it possible to run this project from the root path, without needing to install it.

While running it directly from your clone is acceptable for development reasons (i.e. you are developing `ansible-bundler`), it is a poor choice when it comes to using the software (i.e. developing your playbook), ideally, we should use a packaged version.

For a newly supported OS, I think the best is actually to ship it in a format usable by the OS, and for this specific case, I'm thinking of a homebrew formula. While this is out of the scope, it is the way forward for shipping it for macOS.

l50 committed:
Fixed parameters as per recommendations from author in open PR

l50 committed:
Moved installation instructions earlier in README

- Author wants people to use packages instead of building locally. By moving the installation
	instructions up higher in the README, new users to the project will have an easier time getting
	started.

l50 committed:
Ensure macOS compatibility

l50 committed:
Add homebrew formulae and instructions

l50 committed:
Ran shfmt using provided .editorconfig to ensure formatting is correct

l50 committed:
Execute remaining changes requested by project owner

l50 commented:
Following up to see what else is required to land the PR.

I'm adding some functionality to this project for my own use and figured I'd ship you the stuff that may be helpful.

kriansa commented:
@l50 Hi Jayson,

I apologize for the lack of response, I saw your commits but I didn't have the time to review it yet. I'll do it this weekend.

mattvonrocketstein commented:
I've been testing this also, both for OSX compatibility and because having  a known-working example and usage without system installation would be nice (I can't install the packages on ubuntu, see https://github.com/kriansa/ansible-bundler/issues/15 ).  

For OSX during build I see an issue with non-gnu touch:

```
./bundle-playbook -f ../../../../examples/basic.yaml
Ansible Bundler - v1.10.2

Adding playbook...
Adding ansible config...
Adding entrypoint...
Packaging files to the output...
touch: illegal option -- d
usage:
touch [-A [-][[hh]mm]SS] [-acfhm] [-r file] [-t [[CC]YY]MMDDhhmm[.SS]] file ...
Bundle successfully packaged to ../../../../examples/basic.run
Done.
```

It does generate output anyway.  But afterwards, when trying to run:

```
../../../../examples/basic.run
/tmp/ansible-bundle.5Nicp/run-playbook.sh: line 21: realpath: command not found
Installing playbook Python dependencies...
mkdir: .: No such file or directory
Collecting ansible (from -r /tmp/ansible-bundle.5Nicp/requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/fd/f8/071905c6a67592d0852a9f340f6ab9226861eeeb97fdf4068642b22edcf3/ansible-4.10.0.tar.gz (36.8MB)
    100% |████████████████████████████████| 36.8MB 6.5MB/s
[snip]
Successfully installed MarkupSafe-2.1.3 PyYAML-6.0 ansible ansible-core cffi-1.15.1 cryptography-41.0.1 jinja2-3.1.2 packaging-23.1 pycparser-2.21 resolvelib-0.5.5
You are using pip version 19.0.3, however version 23.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
find: : No such file or directory
/tmp/ansible-bundle.5Nicp/run-playbook.sh: line 114: ansible-playbook: command not found
```
